site/docs/hive.md

```diff
@@ -84,7 +84,32 @@ You should now be able to issue Hive SQL `SELECT` queries using the above table
 SELECT * from table_b;
 ```
 
+#### Using Hadoop Catalog
+Iceberg tables created using `HadoopCatalog` are stored entirely in a directory in a filesytem like HDFS. 
+
+##### Create an Iceberg table
+The first step is to create an Iceberg table using the Spark/Java/Python API and `HadoopCatalog`. For the purposes of this documentation we will assume that the table is called `database_a.table_c` and that the table location is `hdfs://some_path/database_a/table_c`.
+
+##### Create a Hive table
+Now overlay a Hive table on top of this Iceberg table by issuing Hive DDL like so:
+```sql
+CREATE EXTERNAL TABLE table_a 
+STORED BY 'org.apache.iceberg.mr.hive.HiveIcebergStorageHandler' 
+LOCATION 'hdfs://some_bucket/some_path/database_a/table_c';
+```
+
+#### Query the Iceberg table via Hive
+TODO: why does below work if no config settings are set in Hive but fails if we add `set iceberg.mr.catalog=hadoop` like the code suggests we need to do?
```

Comment:
@pvary So for the above, if I create the Hive table but don't set `iceberg.mr.catalog=hadoop` or then I can actually query it successfully. However if I *do* set these (which the code at https://github.com/apache/iceberg/blob/master/mr/src/test/java/org/apache/iceberg/mr/hive/TestTables.java#L133 suggests one should) then I can't query the table. Instead I get this stack trace:
```
2020-11-26T16:46:36,531 INFO  [f94f328b-fe24-4cbc-8723-423fb6ffea6e main([])]: mr.Catalogs (Catalogs.java:loadCatalog(212)) - Loaded Hadoop catalog HadoopCatalog{name=hadoop, location=/tmp/iceberg-test-hadoop-tables}
2020-11-26T16:46:36,533 ERROR [f94f328b-fe24-4cbc-8723-423fb6ffea6e main([])]: hive.log (MetaStoreUtils.java:getDeserializer(456)) - error in initSerDe: org.apache.hadoop.hive.serde2.SerDeException Please provide an existing table or a valid schema
org.apache.hadoop.hive.serde2.SerDeException: Please provide an existing table or a valid schema
        at org.apache.iceberg.mr.hive.HiveIcebergSerDe.initialize(HiveIcebergSerDe.java:62) ~[?:?]
        at org.apache.hadoop.hive.serde2.AbstractSerDe.initialize(AbstractSerDe.java:54) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.serde2.SerDeUtils.initializeSerDe(SerDeUtils.java:533) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.metastore.MetaStoreUtils.getDeserializer(MetaStoreUtils.java:450) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.metastore.MetaStoreUtils.getDeserializer(MetaStoreUtils.java:437) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.metadata.Table.getDeserializerFromMetaStore(Table.java:281) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.metadata.Table.getDeserializer(Table.java:263) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.metadata.Hive.createTable(Hive.java:838) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.metadata.Hive.createTable(Hive.java:872) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.exec.DDLTask.createTable(DDLTask.java:4356) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.exec.DDLTask.execute(DDLTask.java:354) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.exec.Task.executeTask(Task.java:199) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.exec.TaskRunner.runSequential(TaskRunner.java:100) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.Driver.launchTask(Driver.java:2183) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.Driver.execute(Driver.java:1839) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.Driver.runInternal(Driver.java:1526) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1237) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1227) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.cli.CliDriver.processLocalCmd(CliDriver.java:233) ~[hive-cli-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.cli.CliDriver.processCmd(CliDriver.java:184) ~[hive-cli-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.cli.CliDriver.processLine(CliDriver.java:403) ~[hive-cli-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.cli.CliDriver.executeDriver(CliDriver.java:821) ~[hive-cli-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.cli.CliDriver.run(CliDriver.java:759) ~[hive-cli-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.cli.CliDriver.main(CliDriver.java:686) ~[hive-cli-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_252]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_252]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_252]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_252]
        at org.apache.hadoop.util.RunJar.run(RunJar.java:239) ~[hadoop-common-2.8.5-amzn-6.jar:?]
        at org.apache.hadoop.util.RunJar.main(RunJar.java:153) ~[hadoop-common-2.8.5-amzn-6.jar:?]
Caused by: org.apache.iceberg.exceptions.NoSuchTableException: Table does not exist: test.iceberg_table_from_hadoop_catalog
        at org.apache.iceberg.BaseMetastoreCatalog.loadTable(BaseMetastoreCatalog.java:108) ~[?:?]
        at org.apache.iceberg.mr.Catalogs.loadTable(Catalogs.java:100) ~[?:?]
        at org.apache.iceberg.mr.Catalogs.loadTable(Catalogs.java:92) ~[?:?]
        at org.apache.iceberg.mr.hive.HiveIcebergSerDe.initialize(HiveIcebergSerDe.java:60) ~[?:?]
        ... 29 more
2020-11-26T16:46:36,533 ERROR [f94f328b-fe24-4cbc-8723-423fb6ffea6e main([])]: exec.DDLTask (DDLTask.java:failed(639)) - org.apache.hadoop.hive.ql.metadata.HiveException: java.lang.RuntimeException: MetaException(message:org.apache.hadoop.hive.serde2.SerDeException Please provide an existing table or a valid schema)
        at org.apache.hadoop.hive.ql.metadata.Hive.createTable(Hive.java:867)
        at org.apache.hadoop.hive.ql.metadata.Hive.createTable(Hive.java:872)
        at org.apache.hadoop.hive.ql.exec.DDLTask.createTable(DDLTask.java:4356)
        at org.apache.hadoop.hive.ql.exec.DDLTask.execute(DDLTask.java:354)
        at org.apache.hadoop.hive.ql.exec.Task.executeTask(Task.java:199)
        at org.apache.hadoop.hive.ql.exec.TaskRunner.runSequential(TaskRunner.java:100)
        at org.apache.hadoop.hive.ql.Driver.launchTask(Driver.java:2183)
        at org.apache.hadoop.hive.ql.Driver.execute(Driver.java:1839)
        at org.apache.hadoop.hive.ql.Driver.runInternal(Driver.java:1526)
        at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1237)
        at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1227)
        at org.apache.hadoop.hive.cli.CliDriver.processLocalCmd(CliDriver.java:233)
        at org.apache.hadoop.hive.cli.CliDriver.processCmd(CliDriver.java:184)
        at org.apache.hadoop.hive.cli.CliDriver.processLine(CliDriver.java:403)
        at org.apache.hadoop.hive.cli.CliDriver.executeDriver(CliDriver.java:821)
        at org.apache.hadoop.hive.cli.CliDriver.run(CliDriver.java:759)
        at org.apache.hadoop.hive.cli.CliDriver.main(CliDriver.java:686)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.hadoop.util.RunJar.run(RunJar.java:239)
        at org.apache.hadoop.util.RunJar.main(RunJar.java:153)
Caused by: java.lang.RuntimeException: MetaException(message:org.apache.hadoop.hive.serde2.SerDeException Please provide an existing table or a valid schema)
        at org.apache.hadoop.hive.ql.metadata.Table.getDeserializerFromMetaStore(Table.java:283)
        at org.apache.hadoop.hive.ql.metadata.Table.getDeserializer(Table.java:263)
        at org.apache.hadoop.hive.ql.metadata.Hive.createTable(Hive.java:838)
        ... 22 more
Caused by: MetaException(message:org.apache.hadoop.hive.serde2.SerDeException Please provide an existing table or a valid schema)
        at org.apache.hadoop.hive.metastore.MetaStoreUtils.getDeserializer(MetaStoreUtils.java:458)
        at org.apache.hadoop.hive.metastore.MetaStoreUtils.getDeserializer(MetaStoreUtils.java:437)
        at org.apache.hadoop.hive.ql.metadata.Table.getDeserializerFromMetaStore(Table.java:281)
        ... 24 more
```
Any ideas what's going on? I've got a feeling I'm missing something obvious, it's been a long day ;) I'll try debug it tomorrow, just thought you might have some ideas :)

Reply:
Shouldn't the catalog to use also be specified when creating the table? It seems odd that the consumer has to be aware of the catalog used to store the table. Also what if we need to read multiple tables stored in different underlying catalogs?

Reply:
Currently we expect that the catalog is set in the global config (for example: hive-site.xml). Every Iceberg table will use that Catalog. I do not see the exact commands Adrian has executed, but I think Shardul is right and the global config does not contain the catalog settings, and the required values were set between the table creation and the query. Adrian, could you please try:
```
set catalog;
create table;
-- insert data from spark
select from table
```

As for Shardul's question: @lcspinter is started a conversation on the dev list about enabling table level catalog configuration.

Thanks, Peter

Reply:
Thanks @pvary, so what I am doing this is...

First I create the table using the Spark API something along the lines of this:
```
HadoopCatalog catalog = new HadoopCatalog(conf, warehousePath);
TableIdentifier tableId = TableIdentifier.of("hadoop_catalog_db", "iceberg_table_from_hadoop_catalog");
catalog.createTable(tableId, schema, spec);
...
Dataset<Row> ds = spark.createDataFrame(hotels, Hotel.class);
ds.write().format("iceberg").mode("append").save(table.location());
```
This then creates the Iceberg table at `hdfs://host/tmp/iceberg-test-hadoop-tables/hadoop_catalog_db/iceberg_table_from_hadoop_catalog/` - I can see the data and metadata folder on HDFS. So far so good. Then I go to a Hive shell, add the `iceberg-hive-runtime.jar` to the path and then try create a Hive table on top of the above like so:
```
CREATE EXTERNAL TABLE test.iceberg_table_from_hadoop_catalog STORED BY 'org.apache.iceberg.mr.hive.HiveIcebergStorageHandler' LOCATION 'hdfs://host/tmp/iceberg-test-hadoop-tables/hadoop_catalog_db/iceberg_table_from_hadoop_catalog/';
```
This looks like its created OK, I check by doing:
```
show create table test.iceberg_table_from_hadoop_catalog;
OK
CREATE EXTERNAL TABLE `test.iceberg_table_from_hadoop_catalog`(
  `id` bigint COMMENT 'from deserializer', 
  `name` string COMMENT 'from deserializer')
ROW FORMAT SERDE 
  'org.apache.iceberg.mr.hive.HiveIcebergSerDe' 
STORED BY 
  'org.apache.iceberg.mr.hive.HiveIcebergStorageHandler' 
WITH SERDEPROPERTIES ( 
  'serialization.format'='1')
LOCATION
  'hdfs://host/tmp/iceberg-test-hadoop-tables/hadoop_catalog_db/iceberg_table_from_hadoop_catalog'
TBLPROPERTIES (
  'table_type'='ICEBERG', 
  'transient_lastDdlTime'='1606485921')
```
At this point I can successfully perform SELECT statements on the table without setting any conf properties, which surprised me, it seems that this is going via the "Hadoop Tables" path since there is a `location` on the table and it actually works! Then I tried doing the equivalent of what I did in order to read Hive Catalog tables which is to set the value for `iceberg.mr.catalog`:
```
set iceberg.mr.catalog=hadoop;
```
I would expect this to now activate the "Hadoop Catalog" path and the SELECT statement to succeed but instead it fails with the error above. Can you think of anything I'm doing wrong?

Reply:
@shardulm94 I agree with you, at the moment I'm just documenting how it works currently but the idea is for us all to take a step back from the implementation details and look at it through the lens of an end user and make changes. In our earlier implementation we did store the catalog type as a table property so it didn't have to be specified by the end user, I'm not sure how this got lost along the way. I'll test shortly whether this still works or not. I also think we should rename the property to not have `mr` in it. As @pvary says there is some discussion on the mailing list about this and that should lead to us agreeing to make some changes to how this implemented and then updating these docs accordingly.

Reply:
@massdosage: Would it be a problem with the warehouse path for the HadoopCatalog: "location=/tmp/iceberg-test-hadoop-tables"? My guess is that this should be "hdfs://host/tmp/iceberg-test-hadoop-tables", and so the HadoopCatalog is not able to read the table data.

Reply:
OK, I think we can put the `iceberg.mr.catalog` as a table property and this will be used (so it doesn't have to be done at the Hive conf level). I tried this on the HadoopCatalog table above and get the same error when trying to SELECT as when I set the environment variable (which doesn't solve my problem but at least is consistent :) ). @pvary I see the tables automatically created by HiveCatalog don't have that as a table property but perhaps they should if the "hive execution enabled" flag is set to true?

Reply:
> @massdosage: Would it be a problem with the warehouse path for the HadoopCatalog: "location=/tmp/iceberg-test-hadoop-tables"? My guess is that this should be "hdfs://host/tmp/iceberg-test-hadoop-tables", and so the HadoopCatalog is not able to read the table data.

OK, this jogged me into realising that I also need to set "iceberg.mr.catalog.hadoop.warehouse.location". If I do that then it works! Interestingly enough it works with the table location set like you suggested and also with the full path like I had. Anyway, I'll update the document accordingly to show at least how it can be done currently. Thanks!